Validation webhook plugin converts objects to the external version before sending to webhooks

caesarxuchao opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

https://github.com/kubernetes/features/issues/492

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
The apiserver sends external versioned object to the admission webhooks now. Please update the webhooks to expect admissionReview.spec.object.raw to be serialized external versions of objects. 
```

Commit:
let validation webhook convert objects to the external version before sending them

caesarxuchao commented:
/sig api-machinery

caesarxuchao commented:
@deads2k @lavalamp ptal. Thank you.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -196,14 +204,52 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
 	wg := sync.WaitGroup{}
 	wg.Add(len(hooks))
 	for i := range hooks {
+		go func(hook *v1alpha1.ExternalAdmissionHook, i int) {
```
If you pass `i` you don't need to pass `hook`?

lavalamp replied:
Is there a reason to call these in parallel? Seems like a lot of code for possibly small benefit?

caesarxuchao replied:
Removed.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -254,12 +300,12 @@ func (a *GenericAdmissionWebhook) callHook(ctx context.Context, h *v1alpha1.Exte
-	if !matches {
-		return nil
-	}
+	return matches
```
`return false` here; `return true` instead of `break`; delete the `matches` var.

caesarxuchao replied:
Done.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -196,14 +204,52 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
+	wg.Wait()
+	close(hookIndexCh)
+
+	var hookIndice []int
```
Do you need this at all? You can just call len() directly on the channel. If you keep it.

caesarxuchao replied:
Removed.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -196,14 +204,52 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
+	}
+
+	// convert the object to the external version before sending it to the webhook
+	var versioned versionedObjects
```
What if you construct a new attr object instead of making this?

caesarxuchao replied:
Attr is an interface, so if we want to construct a new attr we would have to call `NewAttributesRecord`, and pass it all the `attr.GetObject()`, `attr.GetOldObject()`, etc., which is ugly.

lavalamp replied:
It's a lot easier than that thanks to embedding:

```go
type versioned struct {
  AttributesRecord
  oldObject runtime.Object
  object    runtime.Object
}

func (v versioned) GetObject() ...
func (v versioned) GetOldObject() ...
```

Not claiming you *should* do this, but it would mean 0 changes to the hook call.

caesarxuchao replied:
Good idea. How about revisiting this when I implement the mutating webhooks? Would want to make the hook call as similar as possible.

lavalamp replied:
Sure.

caesarxuchao replied:
Done.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -196,14 +204,52 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
+
+	errCh := make(chan error, len(hookIndice))
+	wg.Add(len(hookIndice))
+	for i := range hookIndice {
```
`_, i` and then you can omit `hookIndice` below. Or switch to ranging over the channel.

caesarxuchao replied:
Right. Anyway , by removing the parallel code, this is gone.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admissionreview.go:
```diff
@@ -54,10 +54,10 @@ func createAdmissionReview(attr admission.Attributes) admissionv1alpha1.Admissio
 			Operation:   admissionv1alpha1.Operation(attr.GetOperation()),
 			Object: runtime.RawExtension{
-				Object: attr.GetObject(),
+				Object: versioned.object,
```
I don't object to the solution in this PR but I think we need to ask ourselves why the objects currently in `attr` are in the internal version. Doesn't that break other things, too?

caesarxuchao replied:
I can see two blockers from building `attr` with versioned objects,
1. defaulting is done as part of the conversion, we don want to run admissions after defaulting
2. the static validation is done on the internal objects.

lavalamp replied:
1. You mean we *do* want to run admission after defaulting?
2. That doesn't mean you have to pass the internal object through the stack?

caesarxuchao replied:
1. yes. fixed the comment.

The current flow and ways to eliminate extra conversion:
* do defaulting on the external types, but it's invoked during the conversion to the internal version.
* convert to external version, run mutating webhooks, convert back to internal version
  * to stop the conversion here, **we need to move the invocation of defaulting out of the conversion stack**
* do validation on internal versions
* convert to external version, run validating webhooks
  * **to stop the conversion here, we need to do validation on external types.**
* convert to the storage version, store to etcd.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -196,14 +204,52 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
+	var versioned versionedObjects
+	gvr := attr.GetResource()
+	if attr.GetOldObject() != nil {
+		versioned.oldObject, err = a.convertor.ConvertToVersion(attr.GetOldObject(), schema.GroupVersion{Group: gvr.Group, Version: gvr.Version})
```
Will this break if we rename a type?

caesarxuchao replied:
Reading the [scheme.go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go), I think if we register the type with both the new and the old name (via `AddKnownTypeWithName`), the conversion will work.

lavalamp replied:
If the webhook is called for the path "apis/indecisive.com/v1/fooname", and realfooname is the internal name, might we end up with "apis/indecisive.com/v1/realfooname" after this call?

caesarxuchao replied:
Changed to use attr.GetKind() to find create a new object, and convert to that object directly.

lavalamp commented on test/images/webhook/Makefile:
```diff
@@ -14,6 +14,7 @@
+	rm -rf webhook
 push:
-	gcloud docker --push gcr.io/kubernetes-e2e-test-images/k8s-sample-admission-webhook-amd64:1.8v1 .
+	gcloud docker -- push gcr.io/kubernetes-e2e-test-images/k8s-sample-admission-webhook-amd64:1.8v2
```
Is the added space in `-- push` intentional?

caesarxuchao replied:
Yes, the original code doesn't work. Awkward :(

lavalamp replied:
Does just plain `gcloud docker push` not work?

caesarxuchao replied:
No. It seems that changed the api: https://cloud.google.com/container-registry/docs/pushing-and-pulling

lavalamp replied:
ack.

lavalamp reviewed (commented):
A few comments.

caesarxuchao commented:
Comments addressed, ptal.

lavalamp commented:
just two comments outstanding still.

caesarxuchao commented:
/retest

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -187,6 +192,32 @@ func (a *GenericAdmissionWebhook) loadConfiguration(attr admission.Attributes) (
 	return hookConfig, nil
 }
 
+type versioned struct {
```
@lavalamp I implemented the embedding.

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -187,6 +192,32 @@ func (a *GenericAdmissionWebhook) loadConfiguration(attr admission.Attributes) (
+	return v.oldObject
+}
+
+func (a *GenericAdmissionWebhook) convertToGVK(obj runtime.Object, gvk schema.GroupVersionKind) (runtime.Object, error) {
```
@lavalamp this should address your concern with type renaming in the same group version.

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -187,6 +192,32 @@ func (a *GenericAdmissionWebhook) loadConfiguration(attr admission.Attributes) (
+	return v.oldObject
+}
+
+func (a *GenericAdmissionWebhook) convertToGVK(obj runtime.Object, gvk schema.GroupVersionKind) (runtime.Object, error) {
```
I'd like to see some unit tests exercising this code that include legacy kinds, kinds across groups, and unstructured/CRD kinds.

@sttts I think this looks pretty good, but I'm concerned about CRDs in particular.  How does it look to you?

caesarxuchao replied:
Sure. The passed in CRD object is of `Unstructured` type, with GVK equal to the `attr.GetKind()`, so this function will be a no-op. I'll add a unit test to confirm.

I think we eventually need an e2e test for webhooks on CRD.

caesarxuchao replied:
Added unit tests as well.

sttts commented:
@deads2k @caesarxuchao taking a look tomorrow.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -187,6 +192,32 @@ func (a *GenericAdmissionWebhook) loadConfiguration(attr admission.Attributes) (
 	return hookConfig, nil
 }
 
+type versioned struct {
```
rename `versionedAttributes`, then the concept is clear.

caesarxuchao replied:
Done.

sttts commented on test/images/webhook/main.go:
```diff
@@ -43,46 +44,56 @@ func (c *Config) addFlags() {
 		}
 	}
+	if !reviewStatus.Allowed {
+		reviewStatus.Result = &metav1.Status{Message: msg}
```
strings.StripSpacesRight(msg)

caesarxuchao replied:
Done.

sttts commented on test/images/webhook/scheme.go:
```diff
@@ -0,0 +1,39 @@
+var codecs = serializer.NewCodecFactory(scheme)
+
+func init() {
+	metav1.AddToGroupVersion(scheme, schema.GroupVersion{Version: "v1"})
```
this is done by `corev1.AddToScheme(scheme)`

caesarxuchao replied:
Removed.

sttts commented on test/images/webhook/main.go:
```diff
@@ -43,46 +44,56 @@ func (c *Config) addFlags() {
-		return nil
+		return toAdmissionReviewStatus(err)
 	}
 	podResource := metav1.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
```
corev1.Resource("pods")

caesarxuchao replied:
Done.

caesarxuchao replied:
Actually can't do this yet, the types don't match: metav1.gvk v.s. schema.gvk.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -196,14 +227,45 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
+		Attributes: attr,
+	}
+	if attr.GetOldObject() != nil {
+		out, err := a.convertToGVK(attr.GetOldObject(), attr.GetKind())
```
// convert to GVK of apiserver endpoint for now.
// TODO: convert to GVK supported by webhook

caesarxuchao replied:
I think "convert to GVK of apiserver endpoint" is what we settled on in the discussion of https://docs.google.com/document/d/1BT8mZaT42jVxtC6l14YMXpUq0vZc6V5MPf_jnzDMMcg/edit#heading=h.5irk4csrpu0y.

sttts replied:
Not sure this is what I get from the document:

> After some discussions here and in the sig-api-machinery meeting, we decided to drop the conversion requirement and let the user specify the exact subject they are interested in. Thus the negotiation protocol we originally discussed here moved to Alternatives Considered section. The only approved section of this design doc is AdmissionResponse type described in the next sub-section.

This sounds like the GVK will be specified in the webhook registration object.

sttts commented:
Looks pretty good. Some small comments. Would be nice to have a simple smoke test that we actually send out versioned objects.

sttts commented:
@deads2k for CRs it will hit our pseudo converter with the right GVK, if I am not mistaken. We don't convert back to internal yet (which would be interesting for CRs). At least this PR doesn't make it worse for CRs. We have to check that CRs' `Unstructured` ends up correctly after serialization.

deads2k commented:
> @deads2k for CRs it will hit our pseudo converter with the right GVK, if I am not mistaken. We don't convert back to internal yet (which would be interesting for CRs). At least this PR doesn't make it worse for CRs. We have to check that CRs' Unstructured ends up correctly after serialization.

I'm not sure that converter is registered with the scheme they're using here.  @caesarxuchao a unit test for it would be very helpful.

caesarxuchao commented:
> I'm not sure that converter is registered with the scheme they're using here

I followed the apiextensio-apiserver start stack, the scheme passed here is the kube-apiserver's Scheme, which doesn't understand CRs. I added a e2e test to make sure the conversion worked. Is it possible to capture it in a unit test? I think the main difficulty is to create a test Scheme that behaves the same as the real Scheme.

lavalamp commented:
We should be running a copy of these for the CRD apiserver, no? And in that copy, the scheme which understands CRDs should be plumbed in.

lavalamp commented:
I would prefer to get this merged and test/fix CRDs in a followup, as getting this merged blocks people building on it, and CRDs can be fixed in parallel.

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/conversion_test.go:
```diff
@@ -0,0 +1,137 @@
+	return s
+}
+
+func TestConvertToGVK(t *testing.T) {
```
@sttts @deads2k PTAL the conversion unit tests.

caesarxuchao commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -68,10 +83,29 @@ var _ = SIGDescribe("AdmissionWebhook", func() {
 		testWebhook(f)
 	})
+
+	It("Should be able to deny custom resource creation", func() {
```
@deads2k @sttts ptal the crd e2e test.

caesarxuchao commented:
@lavalamp @sttts @deads2k updated the code so that it works for CRD as well. Added unit tests and e2e tests for CRD.

I rebased this PR on top of the namespaceSelector one so please ignore the first two commits.

I agree that we should not widen the scope of this PR anymore, otherwise it's too hard for others to work in parallel.

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -187,6 +209,35 @@ func (a *GenericAdmissionWebhook) loadConfiguration(attr admission.Attributes) (
+}
+
+func (a *GenericAdmissionWebhook) convertToGVK(obj runtime.Object, gvk schema.GroupVersionKind) (runtime.Object, error) {
+	if obj.GetObjectKind().GroupVersionKind() == gvk {
```
@sttts @deads2k @lavalamp CR fall into this case and thus will skip the conversion.

sttts replied:
Looks good. On the other hand, I wonder why the convertor does not check the same.

lavalamp replied:
please leave a comment to that effect?

caesarxuchao replied:
>  I wonder why the convertor does not check the same.

Actually without this check, it's the `creator` that fails first. It doesn't know how to create a custom resource. \
I read the conversion stack, I don't see any code specially handles the case where `in` and `out` is of the same type. The defaultConvert will generate the correct result though: https://github.com/kubernetes/kubernetes/blob/7650665059e65b4b22375d1e28da5306536a12fb/staging/src/k8s.io/apimachinery/pkg/conversion/converter.go#L577

caesarxuchao replied:
> please leave a comment to that effect?

Added.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -213,14 +247,49 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
+		}
+		versionedAttr.oldObject = out
+	}
+	if attr.GetObject() != nil {
```
`if obj := attr.GetObject(); obj != nil {` and then reuse obj

caesarxuchao replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/conversion_test.go:
```diff
@@ -0,0 +1,137 @@
+				},
+			},
+		},
+		// TODO: move the replicaset definition to k8s.io/apiserver
```
why do we need another type?

caesarxuchao replied:
I should had removed the todo. It's left there when I used the k8s.io/api's replicaset to prototype the test.

> why do we need another type?

deads2k asked for a test covers cross-group conversion, thus adding the replicaset type to k8s.io/apiserver

caesarxuchao replied:
I named the test so hopefully the intention is obvious now.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/conversion_test.go:
```diff
@@ -0,0 +1,137 @@
+	for _, test := range table {
+		actual, err := w.convertToGVK(test.obj, test.gvk)
+		if err != nil {
+			t.Error(err)
```
print out an index of the test pleaes.

caesarxuchao replied:
Converted to subtests, so that golang's test framework will print out the subtest names.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/conversion_test.go:
```diff
@@ -0,0 +1,137 @@
+			t.Errorf("\nexpected:\n%#v\ngot:\n %#v\n", test.expectedObj, actual)
+		}
+	}
+}
```
I think what I had in mind with my test comment: a test which checks that serialization of the webhook payload correctly serialized the Unstructured object.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/example2/doc.go:
```diff
@@ -0,0 +1,22 @@
+// +k8s:deepcopy-gen=package
+// +groupName=example2.k8s.io
+//
+// package example2 contains an example API used to demonstrate how to create api groups. Moreover, this is
```
nit: example2 API

caesarxuchao replied:
Rewrote the entire description: 
```
// package example2 contains an example API whose internal version is defined in
// another group. THIS IS NOT RECOMMENDED. Unfortunately, Kubernetes have such
// types for historical reasons. This package is for tests.
```

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/example2/install/roundtrip_test.go:
```diff
@@ -0,0 +1,28 @@
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/testing/roundtrip"
+	examplefuzzer "k8s.io/apiserver/pkg/apis/example/fuzzer"
```
example2fuzzer

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/example2/install/install.go:
```diff
@@ -0,0 +1,43 @@
+		&announced.GroupMetaFactoryArgs{
+			GroupName:                  example.GroupName,
+			VersionPreferenceOrder:     []string{example2v1.SchemeGroupVersion.Version},
+			AddInternalObjectsToScheme: example.AddToScheme,
```
Oh, this use the same internal types. This should be reflected in the doc.go description.

caesarxuchao replied:
Done. Sorry.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/admission.go:
```diff
@@ -85,6 +86,7 @@ func (a *AdmissionOptions) ApplyTo(
 	scheme *runtime.Scheme,
 	pluginInitializers ...admission.PluginInitializer,
 ) error {
+	debug.PrintStack()
```
debug code

caesarxuchao replied:
Removed.

sttts commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -105,10 +133,15 @@ func createAuthReaderRoleBinding(f *framework.Framework, namespace string) {
 }
 
 func deployWebhookAndService(f *framework.Framework, image string, context *certContext) {
+
```
nit: no new line

caesarxuchao replied:
Done.

sttts commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -374,3 +409,114 @@ func cleanWebhookTest(f *framework.Framework) {
+	framework.ExpectNoError(err, "registering crd webhook config %s with namespace %s", webhookConfigName, namespace)
+
+	// The webhook configuration is honored in 1s.
+	time.Sleep(2 * time.Second)
```
we really need a condition on the webhook registration telling when it is established. This hurts and asks for flakes.

lavalamp replied:
yes make this 10s to at least have a chance of not having a ton of flakes.

sttts commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -374,3 +409,114 @@ func cleanWebhookTest(f *framework.Framework) {
+}
+
+func testCRDWebhook(f *framework.Framework, crdClient dynamic.ResourceInterface) {
+	By("create a custom resource that should be denied by the webhook")
```
nit: Creating

caesarxuchao replied:
Done

sttts commented on test/images/webhook/main.go:
```diff
@@ -120,7 +122,34 @@ func admitConfigMaps(data []byte) *v1alpha1.AdmissionReviewStatus {
+	}{}
+
+	raw := ar.Spec.Object.Raw
+	err := json.Unmarshal(raw, &cr)
```
about my Unstructured test question: this is e2e and makes sure that Unstructured arrives in the correct format :+1:

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/start.go:
```diff
@@ -102,6 +102,8 @@ func (o CustomResourceDefinitionsServerOptions) Config() (*apiserver.Config, err
 		return nil, err
 	}
 
+	fmt.Printf("CHAO: serverConfig.AdmissionControl=%#v\n", serverConfig.AdmissionControl)
```
Remove debug code :)

caesarxuchao replied:
Sorry. Done.

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go:
```diff
@@ -37,6 +37,26 @@ const (
+// New generates a CRD with random name to avoid name conflict in e2e tests
+func NewNoxuCustomResourceDefinitionForAdmissionWebhookTest(scope apiextensionsv1beta1.ResourceScope) *apiextensionsv1beta1.CustomResourceDefinition {
+	// ensure the singular doesn't end in an s for now
+	gName := names.SimpleNameGenerator.GenerateName("foo") + "a"
```
this looks like a flake waiting to happen.

lavalamp replied:
I thought we were going to do CRD in a second PR. I guess if this works I won't block it.

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go:
```diff
@@ -37,6 +37,26 @@ const (
 )
 
+// New generates a CRD with random name to avoid name conflict in e2e tests
+func NewNoxuCustomResourceDefinitionForAdmissionWebhookTest(scope apiextensionsv1beta1.ResourceScope) *apiextensionsv1beta1.CustomResourceDefinition {
```
Noxu?

sttts replied:
@deads2k's hobby is to invent names and confuse us.

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go:
```diff
@@ -37,6 +37,26 @@ const (
 	noxuInstanceNum int64 = 9223372036854775807
 )
 
+// New generates a CRD with random name to avoid name conflict in e2e tests
```
comment doesn't match function name.

caesarxuchao replied:
Removed the entire function.

caesarxuchao replied:
Removed the function.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/apis/example2/install/install.go:
```diff
@@ -0,0 +1,43 @@
+limitations under the License.
+*/
+
+// Package install installs the certificates API group, making it available as
```
certificates api group?

caesarxuchao replied:
Fixed.

lavalamp commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -283,12 +384,137 @@ func nonCompliantPod(f *framework.Framework) *v1.Pod {
+	_ = client.CoreV1().Namespaces().Delete(skippedNamespaceName, nil)
+}
+
+// NewNoxuCustomResourceDefinitionForAdmissionWebhookTest generates a CRD with random name to avoid name conflict in e2e tests
```
the name doesn't look random?

Noxu?

caesarxuchao replied:
Updated the comment.

caesarxuchao commented:
/retest

caesarxuchao commented:
/retest

lavalamp commented:
/lgtm when you figure out the test problem.

caesarxuchao commented:
Somehow the the order AfterEach function for the `PersistentVolumes GCEPD` test is reversed:

```
I1112 19:04:41.724] [AfterEach] [sig-storage] PersistentVolumes GCEPD
I1112 19:04:41.724]   test/e2e/framework/framework.go:135
I1112 19:04:41.724] Nov 12 19:04:21.604: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
I1112 19:04:41.724] [1mSTEP[0m: Destroying namespace "e2e-tests-pv-p22hk" for this suite.
I1112 19:04:41.724] Nov 12 19:04:41.525: INFO: namespace: e2e-tests-pv-p22hk, resource: bindings, ignored listing per whitelist
I1112 19:04:41.724] Nov 12 19:04:41.692: INFO: namespace e2e-tests-pv-p22hk deletion completed in 19.791344539s
I1112 19:04:41.724] [AfterEach] [sig-storage] PersistentVolumes GCEPD
I1112 19:04:41.725]   test/e2e/storage/persistent_volumes-gce.go:105
I1112 19:04:41.725] Nov 12 19:04:41.692: INFO: AfterEach: Cleaning up test resources
I1112 19:04:41.725] Nov 12 19:04:41.692: INFO: Deleting pod "pvc-tester-p5lgn" in namespace "e2e-tests-pv-p22hk"
I1112 19:04:41.725] Nov 12 19:04:41.714: INFO: Unexpected error occurred: pod Delete API error: Internal error occurred: namespaces "e2e-tests-pv-p22hk" not found
```

caesarxuchao commented:
Actually the order is the same as those tests that passed:
```
I1112 20:44:14.351] [AfterEach] [sig-storage] PersistentVolumes GCEPD
I1112 20:44:14.351]   test/e2e/framework/framework.go:135
I1112 20:44:14.351] Nov 12 20:43:52.362: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
I1112 20:44:14.352] [1mSTEP[0m: Destroying namespace "e2e-tests-pv-2hdrr" for this suite.
I1112 20:44:14.352] Nov 12 20:44:07.420: INFO: namespace: e2e-tests-pv-2hdrr, resource: bindings, ignored listing per whitelist
I1112 20:44:14.352] Nov 12 20:44:07.442: INFO: namespace e2e-tests-pv-2hdrr deletion completed in 15.070836203s
I1112 20:44:14.352] [AfterEach] [sig-storage] PersistentVolumes GCEPD
I1112 20:44:14.352]   test/e2e/storage/persistent_volumes-gce.go:105
I1112 20:44:14.352] Nov 12 20:44:07.442: INFO: AfterEach: Cleaning up test resources
I1112 20:44:14.352] Nov 12 20:44:07.442: INFO: Deleting pod "pvc-tester-tqmm8" in namespace "e2e-tests-pv-2hdrr"
I1112 20:44:14.352] Nov 12 20:44:07.457: INFO: Deleting PersistentVolumeClaim "pvc-nfrwj"
I1112 20:44:14.352] Nov 12 20:44:07.514: INFO: Deleting PersistentVolume "gce-5srbb"
I1112 20:44:14.352] Nov 12 20:44:14.345: INFO: Successfully deleted PD "e2e-5594-fc73409a-c7e9-11e7-9d2a-0a580a3c5ed0".
```

caesarxuchao commented:
I found the problem.

The gce pv test deletes a pod after the namespace is deleted, which is unnecessary:
https://github.com/kubernetes/kubernetes/blob/95038553d7c41ce11ee631e096b773c6756d59a2/test/e2e/storage/persistent_volumes-gce.go#L108

On the other hand, it had been harmless, because if the namespace doesn't exist, apiserver returns 404, which is accepted by the test code.

The webhook e2e test is running in parallel with the gce pv e2e test. The webhook test creates a webhook configuration, requiring all pods CREATE request need to go through the webhook. The gce pv e2e test is **deleting** a pod, not creating. However, the webhook controller checks the namespace exemption before the checks if verb matches:
https://github.com/kubernetes/kubernetes/blob/95038553d7c41ce11ee631e096b773c6756d59a2/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go#L341-L349,

Thus, the DELETE request triggers the webhook controller to retrieve the namespace of the pod: 
https://github.com/kubernetes/kubernetes/blob/95038553d7c41ce11ee631e096b773c6756d59a2/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go#L306. Because the namespace is deleted, the apiserver returns 404. In this PR, I changed the webhook controller's behavior. It now returns a 500 if the namespace doesn't exist, rather than 404. Thus, the pv e2e test failed.

A few action items:
1. in webhook controller, run the rule matching before running the namespace matching, because the former is lighter weight.
2. let the webhook controller return 404 if object's namespace doesn't exist. This keeping apiserver's behavior backwards compatible.
3. remove the unnecessary pod deletion in the pv e2e test.

caesarxuchao commented:
@lavalamp PTAL the second commit. It fixed the e2e test.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/example2/doc.go:
```diff
@@ -0,0 +1,23 @@
+//
+// package example2 contains an example API whose internal version is defined in
+// another group ("example"). THIS IS NOT RECOMMENDED. Unfortunately, Kubernetes
+// have such types for historical reasons. This package is for tests.
```
not only historic types. It's still the prefered pattern when moving a type to another group. Everything else causes huge code duplication.

caesarxuchao replied:
I meant moving types across groups was not recommended. Rephrased to:

```
// package example2 contains an example API whose internal version is defined in
// another group ("example"). This happens if a type is moved to a different
// group. It's not recommended to move types across groups, though Kubernetes
// have a few cases due to historical reasons. This package is for tests.
```

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/example2/install/install.go:
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
nit: 2017

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/example2/install/install.go:
```diff
@@ -0,0 +1,43 @@
+func Install(groupFactoryRegistry announced.APIGroupFactoryRegistry, registry *registered.APIRegistrationManager, scheme *runtime.Scheme) {
+	if err := announced.NewGroupMetaFactory(
+		&announced.GroupMetaFactoryArgs{
+			GroupName:                  example.GroupName,
```
example2.GroupName

sttts commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -376,3 +409,114 @@ func cleanWebhookTest(f *framework.Framework) {
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural:   crdName + "s",
+				Singular: crdName,
+				Kind:     crdName,
```
capital

caesarxuchao replied:
Fixed. The testserver has the same problem: 
https://github.com/kubernetes/kubernetes/blob/81fa823a6ce2dcb95681ff4c0e4bc41ba3c5fd8d/staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go#L49-L52

sttts commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -74,10 +83,29 @@ var _ = SIGDescribe("AdmissionWebhook", func() {
+		deployWebhookAndService(f, "gcr.io/kubernetes-e2e-test-images/k8s-sample-admission-webhook-amd64:1.8v3", context)
+		crdCleanup, dynamicClient := createCRD(f)
+		framework.AddCleanupAction(crdCleanup)
+		registerWebhookForCRD(f, context)
```
Does this mean we leak the webhook? Moreover, our CRD name is constant. What happens if you call the test again? We should at least make sure that this works. If there are flakes, running a test infinitly is usual practice. With the leak this does not work.

caesarxuchao replied:
I'm not sure I completely followed you. Are you asking if we leak the webhook configuration? It's removed in the [cleanWebhookTest](https://github.com/kubernetes/kubernetes/pull/55127/files#diff-fac90cb6856d5c08a931366f5169599cR403). The CRD is removed by `framework.AddCleanupAction(crdCleanup)`. 

I think `AddCleanupAction` is a wrong choice. CleanupAction is called after the entire suite completes. Instead, we should register all the cleanup functions in AfterEach. I'll fix that in another PR.

sttts replied:
Missed `cleanWebhookTest`. Then it's all fine.

janetkuo commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -283,7 +315,7 @@ func registerWebhook(f *framework.Framework, context *certContext) {
 	framework.ExpectNoError(err, "registering webhook config %s with namespace %s", webhookConfigName, namespace)
 
 	// The webhook configuration is honored in 1s.
```
10s

caesarxuchao commented:
/retest

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -394,6 +394,15 @@ func (a *GenericAdmissionWebhook) exemptedByNamespaceSelector(h *v1alpha1.Webhoo
 		return true, nil
 	}
 	namespaceLabels, err := a.getNamespaceLabels(attr)
+	// this means the namespace is not found, for backwards compatibility,
```
Isn't there a namespace checker admission plugin? Maybe it needs to be placed in the list before webhooks?

caesarxuchao replied:
> Isn't there a namespace checker admission plugin? 

The checker always allow DELETE request: https://github.com/kubernetes/kubernetes/blob/1418f9845237840a3e285cad3a763244c4deaaab/staging/src/k8s.io/apiserver/pkg/admission/plugin/namespace/lifecycle/admission.go#L113

Also, we need this webhook to function correctly even if the namespace plugin is turned off.

>  Maybe it needs to be placed in the list before webhooks?

It is :)

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go:
```diff
@@ -406,21 +415,22 @@ func (a *GenericAdmissionWebhook) exemptedByNamespaceSelector(h *v1alpha1.Webhoo
 		m := RuleMatcher{Rule: r, Attr: attr}
 		if m.Matches() {
-			return true, nil
+			matches = true
```
you can add a `break` here, no?

caesarxuchao replied:
Done.

caesarxuchao commented:
Fixed the nits and squashed.

lavalamp commented:
/lgtm

we can fix things in a follow up. I am not convinced that the no namespace error case is correctly handled.

caesarxuchao commented:
I can open another PR to not return an error if it's a DELETE request for a object whose namespace doesn't exist. see https://github.com/kubernetes/kubernetes/issues/49027

lavalamp commented:
looks like this is in the next batch to merge, awesome.
